middle_end/flambda2/to_cmm/to_cmm_primitive.ml

```diff
@@ -692,31 +692,68 @@ let variadic_primitive _env dbg f args =
+       right-to-left evaluation order and does not duplicate any arguments. *)
+    match p with
+    | Nullary _ | Unary _ | Binary _ | Ternary _ -> None
+    | Variadic ((Make_block _ | Make_array _), _) -> Some true
+  in
+  let simple = C.simple ?consider_inlining_effectful_expressions ~dbg in
+  match p with
   | Nullary prim ->
     let extra, expr = nullary_primitive env dbg prim in
     expr, extra, env, res, Ece.pure
-  | Unary (f, x) ->
-    let x, env, eff = C.simple ~dbg env x in
-    let extra, res, expr = unary_primitive env res dbg f x in
+  | Unary (unary, x) ->
+    let x, env, eff = simple env x in
+    let extra, res, expr = unary_primitive env res dbg unary x in
     expr, extra, env, res, eff
-  | Binary (f, x, y) ->
-    let x, env, effx = C.simple ~dbg env x in
-    let y, env, effy = C.simple ~dbg env y in
+  | Binary (binary, x, y) ->
+    let x, env, effx = simple env x in
+    let y, env, effy = simple env y in
     let effs = Ece.join effx effy in
-    let expr = binary_primitive env dbg f x y in
+    let expr = binary_primitive env dbg binary x y in
     expr, None, env, res, effs
-  | Ternary (f, x, y, z) ->
-    let x, env, effx = C.simple ~dbg env x in
-    let y, env, effy = C.simple ~dbg env y in
-    let z, env, effz = C.simple ~dbg env z in
+  | Ternary (ternary, x, y, z) ->
+    let x, env, effx = simple env x in
+    let y, env, effy = simple env y in
+    let z, env, effz = simple env z in
     let effs = Ece.join (Ece.join effx effy) effz in
-    let expr = ternary_primitive env dbg f x y z in
+    let expr = ternary_primitive env dbg ternary x y z in
     expr, None, env, res, effs
-  | Variadic (f, l) ->
-    let args, env, effs = C.simple_list ~dbg env l in
-    let expr = variadic_primitive env dbg f args in
+  | Variadic (((Make_block _ | Make_array _) as variadic), l) ->
+    (* Somewhat counter-intuitively, this correctly matches right-to-left
+       evaluation order---ensuring maximal inlining---since [C.simple_list]
+       translates the first [Simple] in the list first. Consider in pseudo-code:
+
+       let x = <effect-x> in let y = <effect-y> in Make_block [y; x]
+
+       We would like both [x] and [y] to be inlined. The environment will have
+       [y] on the most recent stage since it was the most recent binding. The
+       [Simple] for [y] will be translated first by the code below, meaning
+       inlining is permitted (since [y] is on the most recent stage), producing
+       Make_block [effect-y; y]. Then the [Simple] for [x] will be translated,
+       producing the desired output Make_block [effect-y; effect-x]. The backend
+       will compile this to run effect-x before effect-y by virtue of
+       right-to-left evaluation order. This therefore matches the original code.
+
+       (Same for the binary and ternary cases above, although it usually isn't
+       relevant, as inline-effects-in-cmm is defaulted off.) *)
```

Comment:
Note that this comment also applies to the other primitive translations (where translate primitive arguments left-to-right)

Reply:
I don't understand this comment, can you elaborate?